rust/worker/src/segment/types.rs

```diff
@@ -1,29 +1,362 @@
+#[derive(Debug, Clone)]
 pub(crate) struct MaterializedLogRecord<'a> {
-    pub(super) segment_offset_id: u32,
-    pub(super) log_record: &'a LogRecord,
-    pub(super) materialized_record: DataRecord<'a>,
+    // This is the data record read from the record segment for this id.
+    // None if the record exists only in the log.
+    pub(super) data_record: Option<DataRecord<'a>>,
+    // If present in the record segment then it is the offset id
+    // in the record segment at which the record was found.
+    // If not present in the segment then it is the offset id
+    // at which it should be inserted.
+    pub(super) offset_id: u32,
+    // Set only for the records that are being inserted for the first time
+    // in the log since data_record will be None in such cases. For other
+    // cases, just read from data record.
+    pub(super) user_id: Option<&'a str>,
+    // There can be several entries in the log for an id. This is the final
+    // operation that needs to be done on it. For e.g.
+    // If log has [Update, Update, Delete] then final operation is Delete.
+    // If log has [Insert, Update, Update, Delete] then final operation is Delete.
+    // If log has [Insert, Update, Update] then final operation is Insert.
+    // If log has [Update, Update] then final operation is Update.
+    // All Upserts are converted to Update or Insert as the final operation.
+    // For e.g. if log has [Insert, Upsert] then final operation is update.
+    // If log has [Upsert] and the record does not exist in storage then final
+    // operation is Insert.
+    pub(super) final_operation: Operation,
+    // This is the metadata obtained by combining all the operations
+    // present in the log for this id.
+    // E.g. if has log has [Insert(a: h), Update(a: b, c: d), Update(a: e, f: g)] then this
+    // will contain (a: e, c: d, f: g). This is None if the final operation
+    // above is Delete.
+    pub(super) metadata_to_be_merged: Option<Metadata>,
+    // This is the final document obtained from the last non null operation.
+    // E.g. if log has [Insert(str0), Update(str1), Update(str2), Update()] then this will contain
+    // str2. None if final operation is Delete.
+    pub(super) final_document: Option<&'a str>,
+    // Similar to above, this is the final embedding obtained
+    // from the last non null operation.
+    // E.g. if log has [Insert(emb0), Update(emb1), Update(emb2), Update()]
+    // then this will contain emb2. None if final operation is Delete.
+    pub(super) final_embedding: Option<&'a [f32]>,
 }
 
-impl<'a> MaterializedLogRecord<'a> {
+impl<'a> From<(DataRecord<'a>, u32)> for MaterializedLogRecord<'a> {
+    fn from(data_record_info: (DataRecord<'a>, u32)) -> Self {
+        let data_record = data_record_info.0;
+        let offset_id = data_record_info.1;
+        Self {
+            data_record: Some(data_record),
+            offset_id,
+            user_id: None,
+            final_operation: Operation::Add,
```

Comment:
how come we can just make this add?

Reply:
This is basically the first time you create the materialized record from the corresponding entry in the log, thus it is created in the ADD state and later if there are updates then this state is updated.